invoiceregistry.go

```diff
@@ -732,54 +659,73 @@ func (i *InvoiceRegistry) process(ctx context.Context, h *registryHtlc) error {
 		return nil
 	}
 
-	state, ok := i.invoices[h.rHash]
-	if !ok {
-		// Invoice is not present in memory. Do a db lookup to see if this
-		// happens to be a previously settled invoice.
-		dbInvoice, htlcs, err := i.cdb.Get(ctx, h.rHash)
-		switch {
-		case err == types.ErrInvoiceNotFound:
-			// If the invoice was not found, return a failure
-			// resolution with an invoice not found result.
-			h.resolve(NewFailResolution(ResultInvoiceNotFound))
+	// Re-derive preimage from payment address.
+	mac := hmac.New(sha256.New, i.cfg.PrivKey[:])
 
-			return nil
+	paymentAddr := mpp.PaymentAddr()
+	mac.Write(paymentAddr[:])
+	paymentPreimage, err := lntypes.MakePreimage(mac.Sum(nil))
+	if err != nil {
+		return err
+	}
 
-		case err != nil:
-			return err
-		}
+	// If the preimage doesn't match the payment hash, fail this htlc. Someone
+	// must have tampered with our payment parameters.
+	if paymentPreimage.Hash() != h.rHash {
+		logger.Debugw("Hash mismatch")
 
-		// Fail htlcs paying to unpayable invoices (expired or cancelled).
-		if dbInvoice.State != persistence.InvoiceStateSettleRequested &&
-			dbInvoice.State != persistence.InvoiceStateSettled {
+		h.resolve(NewFailResolution(ResultInvoiceNotFound))
 
-			h.resolve(NewFailResolution(ResultInvoiceNotFound))
+		return nil
+	}
 
-			return nil
-		}
+	// Extract payment parameters.
+	expiry := time.Unix(int64(byteOrder.Uint64(paymentAddr[16:24])), 0)
```

Comment:
I'd rather create a `PaymentAddr` type that receives amt and expiry as construction params and returns them via `Expiry` and `Amount` methods.
This way all the encoding/decoding logics remain on a single place